library/mps/reader.c

```diff
@@ -0,0 +1,632 @@
+/*
+ *  Message Processing Stack, Reader implementation
+ *
+ *  Copyright (C) 2006-2018, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of Mbed TLS (https://tls.mbed.org)
+ */
+
+/* TODO: Use usual include path
+ *       This one is just to make flycheck happy */
+#include "../../include/mbedtls/mps/reader.h"
+#include "../../include/mbedtls/mps/trace.h"
+static int trace_id = TRACE_ID_READER;
+
+#include <string.h>
+
+/*
+ * GENERAL NOTE ON CODING STYLE
+ *
+ * The following code intentionally separates memory loads
+ * and stores from other operations (arithmetic or branches).
+ * This leads to the introduction of many local variables
+ * and significantly increases the C-code line count, but
+ * should not increase the size of generated assembly.
+ *
+ * This reason for this is twofold:
+ * (1) It will ease verification efforts using the VST
+ *     whose program logic cannot directly reason
+ *     about instructions containing a load or store in
+ *     addition to other operations (e.g. *p = *q or
+ *     tmp = *p + 42).
+ * (2) Operating on local variables and writing the results
+ *     back to the target contexts on success only
+ *     allows to maintain structure invariants even
+ *     on failure - this in turn has two benefits:
+ *     (2.a) If for some reason an error code is not caught
+ *           and operation continues, functions are nonetheless
+ *           called with sane contexts, reducing the risk
+ *           of dangerous behavior.
+ *     (2.b) Randomized testing is easier if structures
+ *           remain intact even in the face of failing
+ *           and/or non-sensical calls.
+ *     Moreover, it might even reduce code-size because
+ *     the compiler need not write back temporary results
+ *     to memory in case of failure.
+ *
+ */
+
+int mbedtls_reader_init( mbedtls_reader *rd,
+                         unsigned char *acc, size_t acc_len )
+{
+    mbedtls_reader const zero = { NULL, 0, 0, 0, 0, NULL, 0, 0, { 0 } };
+    *rd = zero;
+    TRACE_INIT( "reader_init, acc len %u", (unsigned) acc_len );
+
+    rd->acc = acc;
+    rd->acc_len = acc_len;
+    RETURN( 0 );
+}
+
+int mbedtls_reader_free( mbedtls_reader *rd )
+{
+    mbedtls_reader const zero = { NULL, 0, 0, 0, 0, NULL, 0, 0, { 0 } };
+    TRACE_INIT( "reader_free" );
+    *rd = zero;
```

Comment:
I think in most (but perhaps not all - I vaguely remember talking about that with @mpg lately?) modules the `xxx_free()` function not only `free()`s heap allocated resources referenced in the given structure, but also zeroizes the latter. As you say, it's arguable whether we want that even if the underlying data is not sensitive, because we could save a few bytes of code-size removing the clearing -- this, however, is a general point we should discuss outside of this PR. 

Here, however, the existing code should be changed in any case, because the present way of zeroization is not reliable, and one should also consider zeroizing the accumulator buffer. I left a note.

Reply:
I think it's better to maintain consistent behavior for now. The matters of security (zeroization) and performance (redundant deinitialization) are too big to be discussed in a comment stream like this.